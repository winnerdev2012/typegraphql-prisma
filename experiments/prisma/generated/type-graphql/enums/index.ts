export { OrderByArg } from "./OrderByArg";
export { PostKind } from "./PostKind";
export { Role } from "./Role";
