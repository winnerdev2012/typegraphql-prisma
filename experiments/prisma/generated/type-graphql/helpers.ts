
export function transformFields(fields: Record<string, any>): Record<string, any> {
  return Object.fromEntries(
    Object.entries(fields)
      // remove __typename and others
      .filter(([key, value]) => !key.startsWith("__"))
      .map<[string, any]>(([key, value]) => {
        if (Object.keys(value).length === 0) {
          return [key, true];
        }
        return [key, transformFields(value)];
      }),
  );
}

