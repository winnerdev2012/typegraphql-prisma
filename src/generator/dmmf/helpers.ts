export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\()+(.+)+(\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\()+(.+)+(\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\()+([A-Za-z])+\:+(.+)+(?:\))+/;

export function parseDocumentationAttributes<TData extends object = object>(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
): Partial<TData> {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return {};
  }
  const rawAttributeArgs = attribute
    ?.match(attributeArgsRegex)?.[0]
    ?.slice(1, -1);
  const splitRawArgsArray = rawAttributeArgs?.split(",").map(it => it.trim());
  const parsedAttributeArgs =
    splitRawArgsArray &&
    (Object.fromEntries(
      splitRawArgsArray.map(it => {
        const [key, value] = it.split(": ");
        return [key, JSON.parse(value)];
      }),
    ) as Partial<TData>);
  return parsedAttributeArgs ?? {};
}
