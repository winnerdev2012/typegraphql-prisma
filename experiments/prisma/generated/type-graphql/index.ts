import { registerEnumType, ObjectType, Field, Int, Float, ID, Resolver, FieldResolver, Root, Ctx, InputType, Query, Mutation, Arg, ArgsType, Args } from "type-graphql";
import DataLoader from "dataloader";

function createGetUserPostsDataLoader(photon: any) {
  const argsToDataLoaderMap = new Map<string, DataLoader<number, Post[] | null>>();
  return function getUserPostsDataLoader(args: any) {
    const argsJSON = JSON.stringify(args);
    let userPostsDataLoader = argsToDataLoaderMap.get(argsJSON);
    if (!userPostsDataLoader) {
      userPostsDataLoader = new DataLoader<number, Post[] | null>(async keys => {
        const fetchedData: any[] = await photon.users.findMany({
          where: { id: { in: keys } },
          select: {
            id: true,
            posts: args,
          },
        });
        return keys
          .map(key => fetchedData.find(data => data.id === key)!)
          .map(data => data.posts);
      });
      argsToDataLoaderMap.set(argsJSON, userPostsDataLoader);
    }
    return userPostsDataLoader;
  }

}

@ArgsType()
export class UserPostsArgs {
  @Field(_type => PostWhereInput, { nullable: true })
  where?: PostWhereInput | null;

  @Field(_type => PostOrderByInput, { nullable: true })
  orderBy?: PostOrderByInput | null;

  @Field(_type => Int, { nullable: true })
  skip?: number | null;

  @Field(_type => ID, { nullable: true })
  after?: string | null;

  @Field(_type => ID, { nullable: true })
  before?: string | null;

  @Field(_type => Int, { nullable: true })
  first?: number | null;

  @Field(_type => Int, { nullable: true })
  last?: number | null;
}

@Resolver(_of => User)
export class UserRelationsResolver {
  @FieldResolver(_type => [Post], {
    nullable: true,
    description: undefined,
  })
  async posts(@Root() user: User, @Ctx() ctx: any, @Args() args: UserPostsArgs): Promise<Post[] | null> {
    ctx.getUserPostsDataLoader = ctx.getUserPostsDataLoader || createGetUserPostsDataLoader(ctx.photon);
    return ctx.getUserPostsDataLoader(args).load(user.id);
  }
}

function createGetPostAuthorDataLoader(photon: any) {
  const argsToDataLoaderMap = new Map<string, DataLoader<string, User>>();
  return function getPostAuthorDataLoader(args: any) {
    const argsJSON = JSON.stringify(args);
    let postAuthorDataLoader = argsToDataLoaderMap.get(argsJSON);
    if (!postAuthorDataLoader) {
      postAuthorDataLoader = new DataLoader<string, User>(async keys => {
        const fetchedData: any[] = await photon.posts.findMany({
          where: { uuid: { in: keys } },
          select: {
            uuid: true,
            author: args,
          },
        });
        return keys
          .map(key => fetchedData.find(data => data.uuid === key)!)
          .map(data => data.author);
      });
      argsToDataLoaderMap.set(argsJSON, postAuthorDataLoader);
    }
    return postAuthorDataLoader;
  }

}

@Resolver(_of => Post)
export class PostRelationsResolver {
  @FieldResolver(_type => User, {
    nullable: false,
    description: undefined,
  })
  async author(@Root() post: Post, @Ctx() ctx: any): Promise<User> {
    ctx.getPostAuthorDataLoader = ctx.getPostAuthorDataLoader || createGetPostAuthorDataLoader(ctx.photon);
    return ctx.getPostAuthorDataLoader({}).load(post.uuid);
  }
}

@ArgsType()
export class FindOneUserArgs {
  @Field(_type => UserWhereUniqueInput, { nullable: false })
  where!: UserWhereUniqueInput;
}

@ArgsType()
export class FindManyUserArgs {
  @Field(_type => UserWhereInput, { nullable: true })
  where?: UserWhereInput | null;

  @Field(_type => UserOrderByInput, { nullable: true })
  orderBy?: UserOrderByInput | null;

  @Field(_type => Int, { nullable: true })
  skip?: number | null;

  @Field(_type => Int, { nullable: true })
  after?: number | null;

  @Field(_type => Int, { nullable: true })
  before?: number | null;

  @Field(_type => Int, { nullable: true })
  first?: number | null;

  @Field(_type => Int, { nullable: true })
  last?: number | null;
}

@ArgsType()
export class CreateOneUserArgs {
  @Field(_type => UserCreateInput, { nullable: false })
  data!: UserCreateInput;
}

@ArgsType()
export class DeleteOneUserArgs {
  @Field(_type => UserWhereUniqueInput, { nullable: false })
  where!: UserWhereUniqueInput;
}

@ArgsType()
export class UpdateOneUserArgs {
  @Field(_type => UserUpdateInput, { nullable: false })
  data!: UserUpdateInput;

  @Field(_type => UserWhereUniqueInput, { nullable: false })
  where!: UserWhereUniqueInput;
}

@ArgsType()
export class UpdateManyUserArgs {
  @Field(_type => UserUpdateManyMutationInput, { nullable: false })
  data!: UserUpdateManyMutationInput;

  @Field(_type => UserWhereInput, { nullable: true })
  where?: UserWhereInput | null;
}

@ArgsType()
export class UpsertOneUserArgs {
  @Field(_type => UserWhereUniqueInput, { nullable: false })
  where!: UserWhereUniqueInput;

  @Field(_type => UserCreateInput, { nullable: false })
  create!: UserCreateInput;

  @Field(_type => UserUpdateInput, { nullable: false })
  update!: UserUpdateInput;
}

@Resolver(_of => User)
export class UserCrudResolver {
  @Query(_returns => User, {
    nullable: true,
    description: undefined
  })
  async findOneUser(@Ctx() ctx: any, @Args() args: FindOneUserArgs): Promise<User | null> {
    return ctx.photon.users.findOne(args);
  }

  @Query(_returns => [User], {
    nullable: true,
    description: undefined
  })
  async findManyUser(@Ctx() ctx: any, @Args() args: FindManyUserArgs): Promise<User[] | null> {
    return ctx.photon.users.findMany(args);
  }

  @Mutation(_returns => User, {
    nullable: false,
    description: undefined
  })
  async createOneUser(@Ctx() ctx: any, @Args() args: CreateOneUserArgs): Promise<User> {
    return ctx.photon.users.create(args);
  }

  @Mutation(_returns => User, {
    nullable: true,
    description: undefined
  })
  async deleteOneUser(@Ctx() ctx: any, @Args() args: DeleteOneUserArgs): Promise<User | null> {
    return ctx.photon.users.delete(args);
  }

  @Mutation(_returns => User, {
    nullable: true,
    description: undefined
  })
  async updateOneUser(@Ctx() ctx: any, @Args() args: UpdateOneUserArgs): Promise<User | null> {
    return ctx.photon.users.update(args);
  }

  @Mutation(_returns => BatchPayload, {
    nullable: false,
    description: undefined
  })
  async updateManyUser(@Ctx() ctx: any, @Args() args: UpdateManyUserArgs): Promise<BatchPayload> {
    return ctx.photon.users.updateMany(args);
  }

  @Mutation(_returns => User, {
    nullable: false,
    description: undefined
  })
  async upsertOneUser(@Ctx() ctx: any, @Args() args: UpsertOneUserArgs): Promise<User> {
    return ctx.photon.users.upsert(args);
  }

  @Query(_returns => AggregateUser, {
    nullable: false,
    description: undefined
  })
  async aggregateUser(@Ctx() ctx: any): Promise<AggregateUser> {
    return ctx.photon.users.aggregate();
  }
}

@ArgsType()
export class FindOnePostArgs {
  @Field(_type => PostWhereUniqueInput, { nullable: false })
  where!: PostWhereUniqueInput;
}

@ArgsType()
export class FindManyPostArgs {
  @Field(_type => PostWhereInput, { nullable: true })
  where?: PostWhereInput | null;

  @Field(_type => PostOrderByInput, { nullable: true })
  orderBy?: PostOrderByInput | null;

  @Field(_type => Int, { nullable: true })
  skip?: number | null;

  @Field(_type => ID, { nullable: true })
  after?: string | null;

  @Field(_type => ID, { nullable: true })
  before?: string | null;

  @Field(_type => Int, { nullable: true })
  first?: number | null;

  @Field(_type => Int, { nullable: true })
  last?: number | null;
}

@ArgsType()
export class CreateOnePostArgs {
  @Field(_type => PostCreateInput, { nullable: false })
  data!: PostCreateInput;
}

@ArgsType()
export class DeleteOnePostArgs {
  @Field(_type => PostWhereUniqueInput, { nullable: false })
  where!: PostWhereUniqueInput;
}

@ArgsType()
export class UpdateOnePostArgs {
  @Field(_type => PostUpdateInput, { nullable: false })
  data!: PostUpdateInput;

  @Field(_type => PostWhereUniqueInput, { nullable: false })
  where!: PostWhereUniqueInput;
}

@ArgsType()
export class UpdateManyPostArgs {
  @Field(_type => PostUpdateManyMutationInput, { nullable: false })
  data!: PostUpdateManyMutationInput;

  @Field(_type => PostWhereInput, { nullable: true })
  where?: PostWhereInput | null;
}

@ArgsType()
export class UpsertOnePostArgs {
  @Field(_type => PostWhereUniqueInput, { nullable: false })
  where!: PostWhereUniqueInput;

  @Field(_type => PostCreateInput, { nullable: false })
  create!: PostCreateInput;

  @Field(_type => PostUpdateInput, { nullable: false })
  update!: PostUpdateInput;
}

@Resolver(_of => Post)
export class PostCrudResolver {
  @Query(_returns => Post, {
    nullable: true,
    description: undefined
  })
  async findOnePost(@Ctx() ctx: any, @Args() args: FindOnePostArgs): Promise<Post | null> {
    return ctx.photon.posts.findOne(args);
  }

  @Query(_returns => [Post], {
    nullable: true,
    description: undefined
  })
  async findManyPost(@Ctx() ctx: any, @Args() args: FindManyPostArgs): Promise<Post[] | null> {
    return ctx.photon.posts.findMany(args);
  }

  @Mutation(_returns => Post, {
    nullable: false,
    description: undefined
  })
  async createOnePost(@Ctx() ctx: any, @Args() args: CreateOnePostArgs): Promise<Post> {
    return ctx.photon.posts.create(args);
  }

  @Mutation(_returns => Post, {
    nullable: true,
    description: undefined
  })
  async deleteOnePost(@Ctx() ctx: any, @Args() args: DeleteOnePostArgs): Promise<Post | null> {
    return ctx.photon.posts.delete(args);
  }

  @Mutation(_returns => Post, {
    nullable: true,
    description: undefined
  })
  async updateOnePost(@Ctx() ctx: any, @Args() args: UpdateOnePostArgs): Promise<Post | null> {
    return ctx.photon.posts.update(args);
  }

  @Mutation(_returns => BatchPayload, {
    nullable: false,
    description: undefined
  })
  async updateManyPost(@Ctx() ctx: any, @Args() args: UpdateManyPostArgs): Promise<BatchPayload> {
    return ctx.photon.posts.updateMany(args);
  }

  @Mutation(_returns => Post, {
    nullable: false,
    description: undefined
  })
  async upsertOnePost(@Ctx() ctx: any, @Args() args: UpsertOnePostArgs): Promise<Post> {
    return ctx.photon.posts.upsert(args);
  }

  @Query(_returns => AggregatePost, {
    nullable: false,
    description: undefined
  })
  async aggregatePost(@Ctx() ctx: any): Promise<AggregatePost> {
    return ctx.photon.posts.aggregate();
  }
}
