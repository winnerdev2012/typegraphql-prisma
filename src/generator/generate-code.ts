import { DMMF } from "@prisma/client/runtime";
import { Project } from "ts-morph";
import path from "path";

import { noop } from "./helpers";
import generateEnumFromDef from "./enum";
import generateObjectTypeClassFromModel from "./object-type-class";
import generateRelationsResolverClassesFromModel from "./resolvers/relations";
import {
  generateOutputTypeClassFromType,
  generateInputTypeClassFromType,
} from "./type-class";
import generateCrudResolverClassFromMapping from "./resolvers/full-crud";
import {
  resolversFolderName,
  relationsResolversFolderName,
  crudResolversFolderName,
  inputsFolderName,
  outputsFolderName,
  enumsFolderName,
  modelsFolderName,
} from "./config";
import {
  generateResolversBarrelFile,
  generateInputsBarrelFile,
  generateOutputsBarrelFile,
  generateIndexFile,
  generateModelsBarrelFile,
  generateEnumsBarrelFile,
} from "./imports";
import saveSourceFile from "../utils/saveSourceFile";
import { GenerateCodeOptions } from "./options";

export default async function generateCode(
  dmmf: DMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop,
) {
  const baseDirPath = options.outputDirPath;
  const project = new Project();
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map(model => model.name);

  log("Generating enums...");
  const datamodelEnumNames = dmmf.datamodel.enums.map(enumDef => enumDef.name);
  await Promise.all(
    dmmf.datamodel.enums.map(enumDef =>
      generateEnumFromDef(project, baseDirPath, enumDef),
    ),
  );
  await Promise.all(
    dmmf.schema.enums
      // skip enums from datamodel
      .filter(enumDef => !datamodelEnumNames.includes(enumDef.name))
      .map(enumDef => generateEnumFromDef(project, baseDirPath, enumDef)),
  );
  const emittedEnumNames = [
    ...new Set([
      ...dmmf.schema.enums.map(it => it.name),
      ...dmmf.datamodel.enums.map(it => it.name),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, "index.ts"),
    undefined,
    { overwrite: true },
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);
  await saveSourceFile(enumsBarrelExportSourceFile);

  log("Generating models...");
  await Promise.all(
    dmmf.datamodel.models.map(model =>
      generateObjectTypeClassFromModel(project, baseDirPath, model, modelNames),
    ),
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, "index.ts"),
    undefined,
    { overwrite: true },
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmf.datamodel.models.map(it => it.name),
  );
  await saveSourceFile(modelsBarrelExportSourceFile);

  log("Generating output types...");
  const rootTypes = dmmf.schema.outputTypes.filter(type =>
    ["Query", "Mutation"].includes(type.name),
  );
  const outputTypesToGenerate = dmmf.schema.outputTypes.filter(
    // skip generating models and root resolvers
    type => !modelNames.includes(type.name) && !rootTypes.includes(type),
  );
  await Promise.all(
    outputTypesToGenerate.map(type =>
      generateOutputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        modelNames,
      ),
    ),
  );
  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map(it => it.name),
  );
  await saveSourceFile(outputsBarrelExportSourceFile);

  log("Generating input types...");
  await Promise.all(
    dmmf.schema.inputTypes.map(type =>
      generateInputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        modelNames,
      ),
    ),
  );
  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmf.schema.inputTypes.map(it => it.name),
  );
  await saveSourceFile(inputsBarrelExportSourceFile);

  log("Generating relation resolvers...");
  const relationResolversData = await Promise.all(
    dmmf.datamodel.models
      .filter(model => model.fields.some(field => field.relationName))
      .map(model => {
        const outputType = dmmf.schema.outputTypes.find(
          type => type.name === model.name,
        )!;
        const mapping = dmmf.mappings.find(it => it.model === model.name)!;
        return generateRelationsResolverClassesFromModel(
          project,
          baseDirPath,
          model,
          mapping,
          outputType,
          modelNames,
        );
      }),
  );
  if (relationResolversData.length > 0) {
    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        "index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateResolversBarrelFile(
      "relations",
      relationResolversBarrelExportSourceFile,
      relationResolversData,
    );
    await saveSourceFile(relationResolversBarrelExportSourceFile);
  }

  log("Generating crud resolvers...");
  const crudResolversData = await Promise.all(
    dmmf.mappings.map(mapping =>
      generateCrudResolverClassFromMapping(
        project,
        baseDirPath,
        mapping,
        rootTypes,
        modelNames,
        options,
      ),
    ),
  );
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateResolversBarrelFile(
    "crud",
    crudResolversBarrelExportSourceFile,
    crudResolversData,
  );
  await saveSourceFile(crudResolversBarrelExportSourceFile);

  log("Generating index file");
  const indexSourceFile = project.createSourceFile(
    baseDirPath + "/index.ts",
    undefined,
    { overwrite: true },
  );
  generateIndexFile(indexSourceFile, relationResolversData.length > 0);
  await saveSourceFile(indexSourceFile);
}
