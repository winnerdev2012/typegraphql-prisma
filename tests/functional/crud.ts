import "reflect-metadata";
import { promises as fs } from "fs";
import { buildSchema } from "type-graphql";
import { graphql, GraphQLSchema } from "graphql";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { generateCodeFromSchema } from "../helpers/generate-code";

describe("crud resolvers execution", () => {
  let outputDirPath: string;
  let graphQLSchema: GraphQLSchema;

  beforeAll(async () => {
    outputDirPath = generateArtifactsDirPath("functional-crud");
    await fs.mkdir(outputDirPath, { recursive: true });
    const prismaSchema = /* prisma */ `
      model User {
        intIdField          Int     @id @default(autoincrement())
        uniqueStringField   String  @unique
        optionalStringField String?
        dateField           DateTime
      }
    `;
    await generateCodeFromSchema(prismaSchema, { outputDirPath });
    const { UserCrudResolver } = require(outputDirPath +
      "/resolvers/crud/User/UserCrudResolver.ts");

    graphQLSchema = await buildSchema({
      resolvers: [UserCrudResolver],
      validate: false,
    });
  });

  it("should properly call PrismaClient on `findOne` action", async () => {
    const document = /* graphql */ `
      query {
        user(where: { uniqueStringField: "uniqueValue" }) {
          intIdField
          dateField
        }
      }
    `;
    const prismaMock = {
      user: {
        findOne: jest.fn().mockResolvedValue({
          intIdField: 1,
          dateField: new Date("2019-12-31T14:16:02.572Z"),
        }),
      },
    };

    const { data, errors } = await graphql(graphQLSchema, document, null, {
      prisma: prismaMock,
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("user mocked response");
    expect(prismaMock.user.findOne.mock.calls).toMatchSnapshot(
      "findOneUser call args",
    );
  });

  it("should properly call PrismaClient on `findMany` action", async () => {
    const document = /* graphql */ `
      query {
        users(
          take: 1
          skip: 1
          orderBy: { intIdField: desc }
          where: { dateField: { lte: "2019-12-31T19:16:02.572Z" } }
        ) {
          intIdField
          dateField
        }
      }
    `;
    const prismaMock = {
      user: {
        findMany: jest.fn().mockResolvedValue([
          {
            intIdField: 1,
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          },
        ]),
      },
    };

    const { data, errors } = await graphql(graphQLSchema, document, null, {
      prisma: prismaMock,
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("users mocked response");
    expect(prismaMock.user.findMany.mock.calls).toMatchSnapshot(
      "findManyUser call args",
    );
  });

  it("should properly call PrismaClient on `create` action", async () => {
    const document = /* graphql */ `
      mutation {
        createUser(
          data: {
            uniqueStringField: "unique"
            optionalStringField: "optional"
            dateField: "2019-12-31T14:16:02.572Z"
          }
        ) {
          intIdField
          dateField
        }
      }
    `;
    const prismaMock = {
      user: {
        create: jest.fn().mockResolvedValue({
          intIdField: 1,
          dateField: new Date("2019-12-31T14:16:02.572Z"),
        }),
      },
    };

    const { data, errors } = await graphql(graphQLSchema, document, null, {
      prisma: prismaMock,
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("createUser mocked response");
    expect(prismaMock.user.create.mock.calls).toMatchSnapshot(
      "createUser call args",
    );
  });

  it("should properly call PrismaClient on `delete` action", async () => {
    const document = /* graphql */ `
      mutation {
        deleteUser(
          where: {
            uniqueStringField: "unique"
          }
        ) {
          intIdField
          dateField
        }
      }
    `;
    const prismaMock = {
      user: {
        delete: jest.fn().mockResolvedValue({
          intIdField: 1,
          dateField: new Date("2019-12-31T14:16:02.572Z"),
        }),
      },
    };

    const { data, errors } = await graphql(graphQLSchema, document, null, {
      prisma: prismaMock,
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("deleteUser mocked response");
    expect(prismaMock.user.delete.mock.calls).toMatchSnapshot(
      "deleteUser call args",
    );
  });

  it("should properly call PrismaClient on `update` action", async () => {
    const document = /* graphql */ `
      mutation {
        updateUser(
          data: {
            dateField: "2019-12-31T14:16:02.572Z",
          }
          where: {
            uniqueStringField: "unique"
          }
        ) {
          intIdField
          dateField
        }
      }
    `;
    const prismaMock = {
      user: {
        update: jest.fn().mockResolvedValue({
          intIdField: 1,
          dateField: new Date("2019-12-31T14:16:02.572Z"),
        }),
      },
    };

    const { data, errors } = await graphql(graphQLSchema, document, null, {
      prisma: prismaMock,
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("updateUser mocked response");
    expect(prismaMock.user.update.mock.calls).toMatchSnapshot(
      "updateUser call args",
    );
  });

  it("should properly call PrismaClient on `updateMany` action", async () => {
    const document = /* graphql */ `
      mutation {
        updateManyUser(
          data: {
            optionalStringField: null,
          }
          where: {
            dateField: { lte: "2019-12-31T19:16:02.572Z" }
          }
        ) {
          count
        }
      }
    `;
    const prismaMock = {
      user: {
        updateMany: jest.fn().mockResolvedValue({
          count: 3,
        }),
      },
    };

    const { data, errors } = await graphql(graphQLSchema, document, null, {
      prisma: prismaMock,
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("updateManyUser mocked response");
    expect(prismaMock.user.updateMany.mock.calls).toMatchSnapshot(
      "updateManyUser call args",
    );
  });

  it("should properly call PrismaClient on `deleteMany` action", async () => {
    const document = /* graphql */ `
      mutation {
        deleteManyUser(
          where: {
            dateField: { lte: "2019-12-31T19:16:02.572Z" }
          }
        ) {
          count
        }
      }
    `;
    const prismaMock = {
      user: {
        deleteMany: jest.fn().mockResolvedValue({
          count: 3,
        }),
      },
    };

    const { data, errors } = await graphql(graphQLSchema, document, null, {
      prisma: prismaMock,
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("deleteManyUser mocked response");
    expect(prismaMock.user.deleteMany.mock.calls).toMatchSnapshot(
      "deleteManyUser call args",
    );
  });

  it("should properly call PrismaClient on `upsert` action", async () => {
    const document = /* graphql */ `
      mutation {
        upsertUser(
          where: {
            uniqueStringField: "unique"
          }
          create: {
            uniqueStringField: "unique"
            optionalStringField: "optional"
            dateField: "2019-12-31T14:16:02.572Z"
          }
          update: {
            optionalStringField: null,
          }
        ) {
          intIdField
          uniqueStringField
          optionalStringField
          dateField
        }
      }
    `;
    const prismaMock = {
      user: {
        upsert: jest.fn().mockResolvedValue({
          intIdField: 1,
          uniqueStringField: "unique",
          optionalStringField: "optional",
          dateField: new Date("2019-12-31T14:16:02.572Z"),
        }),
      },
    };

    const { data, errors } = await graphql(graphQLSchema, document, null, {
      prisma: prismaMock,
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("upsertUser mocked response");
    expect(prismaMock.user.upsert.mock.calls).toMatchSnapshot(
      "upsertUser call args",
    );
  });

  it("should properly call PrismaClient on `aggregate` action", async () => {
    const document = /* graphql */ `
      query {
        aggregateUser {
          count(
            take: 1
            skip: 1
            orderBy: { intIdField: desc }
            where: { dateField: { lte: "2019-12-31T19:16:02.572Z" } }
          )
        }
      }
    `;
    const prismaMock = {
      user: {
        count: jest.fn().mockResolvedValue(5),
      },
    };

    const { data, errors } = await graphql(graphQLSchema, document, null, {
      prisma: prismaMock,
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("aggregateUserCount mocked response");
    expect(prismaMock.user.count.mock.calls).toMatchSnapshot(
      "aggregateUserCount call args",
    );
  });
});
