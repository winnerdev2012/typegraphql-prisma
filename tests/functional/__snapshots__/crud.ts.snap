// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crud resolvers execution should properly call PrismaClient on \`aggregate\` action: aggregateUserCount call args 1`] = `
Array [
  Array [
    AggregateUserCountArgs {
      "orderBy": UserOrderByInput {
        "intIdField": "desc",
      },
      "skip": 1,
      "take": 1,
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution should properly call PrismaClient on \`aggregate\` action: aggregateUserCount mocked response 1`] = `
Object {
  "aggregateUser": Object {
    "count": 5,
  },
}
`;

exports[`crud resolvers execution should properly call PrismaClient on \`create\` action: createUser call args 1`] = `
Array [
  Array [
    CreateUserArgs {
      "data": UserCreateInput {
        "dateField": 2019-12-31T14:16:02.572Z,
        "optionalStringField": "optional",
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution should properly call PrismaClient on \`create\` action: createUser mocked response 1`] = `
Object {
  "createUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution should properly call PrismaClient on \`delete\` action: deleteUser call args 1`] = `
Array [
  Array [
    DeleteUserArgs {
      "where": UserWhereUniqueInput {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution should properly call PrismaClient on \`delete\` action: deleteUser mocked response 1`] = `
Object {
  "deleteUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution should properly call PrismaClient on \`deleteMany\` action: deleteManyUser call args 1`] = `
Array [
  Array [
    DeleteManyUserArgs {
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution should properly call PrismaClient on \`deleteMany\` action: deleteManyUser mocked response 1`] = `
Object {
  "deleteManyUser": Object {
    "count": 3,
  },
}
`;

exports[`crud resolvers execution should properly call PrismaClient on \`findMany\` action: findManyUser call args 1`] = `
Array [
  Array [
    FindManyUserArgs {
      "orderBy": UserOrderByInput {
        "intIdField": "desc",
      },
      "skip": 1,
      "take": 1,
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution should properly call PrismaClient on \`findMany\` action: users mocked response 1`] = `
Object {
  "users": Array [
    Object {
      "dateField": "2019-12-31T14:16:02.572Z",
      "intIdField": 1,
    },
  ],
}
`;

exports[`crud resolvers execution should properly call PrismaClient on \`findOne\` action: findOneUser call args 1`] = `
Array [
  Array [
    FindOneUserArgs {
      "where": UserWhereUniqueInput {
        "uniqueStringField": "uniqueValue",
      },
    },
  ],
]
`;

exports[`crud resolvers execution should properly call PrismaClient on \`findOne\` action: user mocked response 1`] = `
Object {
  "user": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution should properly call PrismaClient on \`update\` action: updateUser call args 1`] = `
Array [
  Array [
    UpdateUserArgs {
      "data": UserUpdateInput {
        "dateField": 2019-12-31T14:16:02.572Z,
      },
      "where": UserWhereUniqueInput {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution should properly call PrismaClient on \`update\` action: updateUser mocked response 1`] = `
Object {
  "updateUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution should properly call PrismaClient on \`updateMany\` action: updateManyUser call args 1`] = `
Array [
  Array [
    UpdateManyUserArgs {
      "data": UserUpdateManyMutationInput {
        "optionalStringField": null,
      },
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution should properly call PrismaClient on \`updateMany\` action: updateManyUser mocked response 1`] = `
Object {
  "updateManyUser": Object {
    "count": 3,
  },
}
`;

exports[`crud resolvers execution should properly call PrismaClient on \`upsert\` action: upsertUser call args 1`] = `
Array [
  Array [
    UpsertUserArgs {
      "create": UserCreateInput {
        "dateField": 2019-12-31T14:16:02.572Z,
        "optionalStringField": "optional",
        "uniqueStringField": "unique",
      },
      "update": UserUpdateInput {
        "optionalStringField": null,
      },
      "where": UserWhereUniqueInput {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution should properly call PrismaClient on \`upsert\` action: upsertUser mocked response 1`] = `
Object {
  "upsertUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
    "optionalStringField": "optional",
    "uniqueStringField": "unique",
  },
}
`;
